### PR TITLE
Add maximumCapacity to taskRunner

### PR DESCRIPTION
**Pull request by georgew5656**
Move logic for calculating maximumCapacity (tatal capacity based on max workers from autoscaling) to task runners
### Description
It is possible to get weird responses from the /totalWorkerCapacity endpoint if mmless ingestion is enabled and the overlord dynamic.autoscaler config is set. This is because the TaskQueryTool.getTotalWorkerCapacity gets totalCapacity from the overlord's task runner but gets maximumCapacity directly from the dynamic config.

I think it makes sense to just expose a getMaximumCapacity method on TaskRunner that defaults to -1 (the default value of maximumCapacity) and gets overwritten.

#### Added a forbidden-apis entry ...
In order to move getMaximumCapacity logic into each TaskRunner, i had to duplicate logic between HttpRemoteTaskRunner and RemoteTaskRunner since they have the same logic (check the overlord dynamic configuration), but I think this is okay in order for the task runners to be responsible for their own max capacity values.

#### Release note
- TaskRunner should expose a getMaximumCapacity field

##### Key changed/added classes in this PR
 * `TaskRunner`
 * `HttpRemoteTaskRunner/RemoteTaskRunner`
 * `KubernetesAndWorkerTaskRunner/KubernetesTaskRunner`
 * `TaskQueryTool`

This PR has:

- [X] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java [hunk 1]
```diff
@@ -155,6 +155,15 @@ default int getTotalCapacity()
     return -1;
   }
 
+  /**
+   * The maximum number of tasks this TaskRunner can run concurrently with autoscaling hints.
+   * Can return -1 if this method is not implemented or capacity can't be found.
+   */
+  default int getMaximumCapacity()
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java:162]:
```suggestion
  default int getMaximumCapacityWithAutoScale()
```

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java [hunk 2]
```diff
@@ -155,6 +155,15 @@ default int getTotalCapacity()
     return -1;
   }
 
+  /**
+   * The maximum number of tasks this TaskRunner can run concurrently with autoscaling hints.
+   * Can return -1 if this method is not implemented or capacity can't be found.
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java:160]:
```suggestion
   * @return -1 if this method is not implemented or capacity can't be found.
```

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/RemoteTaskRunner.java [hunk 3]
```diff
@@ -1648,6 +1649,30 @@ public int getTotalCapacity()
     return getWorkers().stream().mapToInt(workerInfo -> workerInfo.getWorker().getCapacity()).sum();
   }
 
+  @Override
+  public int getMaximumCapacity()
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/RemoteTaskRunner.java:1653]: Maybe add a javadoc here listing the cases when this method returns -1.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/hrtr/HttpRemoteTaskRunner.java [hunk 4]
```diff
@@ -1791,6 +1792,30 @@ public int getTotalCapacity()
     return getWorkers().stream().mapToInt(workerInfo -> workerInfo.getWorker().getCapacity()).sum();
   }
 
+  @Override
+  public int getMaximumCapacity()
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/hrtr/HttpRemoteTaskRunner.java:1796]: Add a javadoc here listing the cases when this method returns -1.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueryTool.java [hunk 5]
```diff
@@ -377,38 +374,9 @@ public TotalWorkerCapacityResponse getTotalWorkerCapacity()
...
-    } else {
-      // Auto-scale is not using DefaultWorkerBehaviorConfig
-      log.debug(
-          "Cannot calculate maximum worker capacity as WorkerBehaviorConfig [%s] of type [%s] does not support getting max capacity",
-          workerBehaviorConfig,
-          workerBehaviorConfig.getClass().getSimpleName()
-      );
-      maximumCapacity = -1;
-    }
+    int maximumCapacity = taskRunner.getMaximumCapacity();
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueryTool.java:379]: Maybe just inline these in the constructor call.

**Review by kfaraz — COMMENTED** [review]
Overall looks good, left some minor suggestions.

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesAndWorkerTaskRunner.java [hunk 6]
```diff
@@ -268,6 +268,12 @@ public int getTotalCapacity()
     return Math.max(0, k8sCapacity) + Math.max(0, workerCapacity);
   }
 
+  @Override
+  public int getMaximumCapacity()
+  {
+    return workerTaskRunner.getMaximumCapacity() + kubernetesTaskRunner.getMaximumCapacity();
```
Comment (suneet-s) [inline extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesAndWorkerTaskRunner.java:274]: The response for max capacity should depend on the underlying runner strategy.

Comment (georgew5656) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesAndWorkerTaskRunner.java:274]: i can update this response in a separate PR, i think its not so simple to calculate what the capacity should be

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/RemoteTaskRunner.java (hunk 3 above)
Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/RemoteTaskRunner.java:1653]: I wonder if this method body should be commoned out between `RemoteTaskRunner` and `HttpTaskRunner`.

Comment (georgew5656) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/RemoteTaskRunner.java]: were you thinking of introducing a new class to do this?

**Review by kfaraz — APPROVED** [review]

**Pull request merged**